In [6]:
# Récuperation des données

In [2]:
function readKnaptxtInstance(filename)
    price=[]
    weight=[]
    KnapCap=[]
    open(filename) do f
        for i in 1:3
            tok = split(readline(f))
            if(tok[1] == "ListPrices=")
                for i in 2:(length(tok)-1)
                    push!(price,parse(Int64, tok[i]))
                end
            elseif(tok[1] == "ListWeights=")
                for i in 2:(length(tok)-1)
                    push!(weight,parse(Int64, tok[i]))
                end
            elseif(tok[1] == "Capacity=")
                push!(KnapCap, parse(Int64, tok[2]))
            else
                println("Unknown read :", tok)
            end 
        end
    end
    capacity=KnapCap[1]
    return price, weight, capacity
end

readKnaptxtInstance (generic function with 1 method)

In [3]:
function creerTab(price,weight,capacity)
    n = length(price)
    m = capacity
    # on initialise un tableau contenant la valeur max et l'historique des objets pris
    C = [(0, []) for i in 1:n, j in 1:(m+1)]
    # on remplit la première ligne
    for j in 2:(m+1)
        # on regarde si le poids de l'objet et plus petit que notre poids maximal permis j
        # si oui, on ajoute l'objet 1 à l'historique ainsi que sa valeur
        if (j > weight[1])
            C[1,j] = (price[1], [1])
        end
    end
    #on remplit les autres lignes
    for i in 2:n
        for j in 2:(m+1)
            # on regarde si C[i, j - Wi] existe
            # si non, on prend la valeur du dessus
            if (j-weight[i] <= 0)
                C[i,j] = C[i-1,j]
            else
                # si oui, on applique la relation de récurrence
                if C[i-1,j][1] >= C[i-1,j-weight[i]][1]+price[i]
                    C[i,j] = C[i-1,j]
                else
                    l = copy(C[i-1,j-weight[i]][2])
                    C[i,j] = (C[i-1,j-weight[i]][1]+price[i], push!(l,i))
                end
            end
        end
    end
    resultat = C
    return resultat
end

creerTab (generic function with 1 method)

On remarque que les objets choisis sont ceux ayant le rapport valeur/poids le plus élévé et dont la somme des poids est proche de notre capacité. Ce qui semble être une bonne solution à notre problème.

In [4]:
prix,poids,cap = readKnaptxtInstance("instancesETU/KNAPnewformat/test.opb.txt")
r = creerTab(prix,poids,cap)

4×11 Matrix{Tuple{Int64, Vector{Any}}}:
 (0, [])  (0, [])  (0, [])  (0, [])    …  (42, [1])     (42, [1])
 (0, [])  (0, [])  (0, [])  (0, [])       (42, [1])     (42, [1])
 (0, [])  (0, [])  (0, [])  (12, [3])     (52, [2, 3])  (54, [1, 3])
 (0, [])  (0, [])  (0, [])  (12, [3])     (65, [2, 4])  (65, [2, 4])

La valeur maximal du sac à dos est de 65 pour une capacité de 10, avec les objets 2 et 4

In [5]:
prix,poids,cap = readKnaptxtInstance("instancesETU/KNAPnewformat/almost_strongly_correlated/knapPI_5_100_10000_1_-23965.opb.txt")
r = creerTab(prix,poids,cap)

100×9985 Matrix{Tuple{Int64, Vector{Any}}}:
 (0, [])  …  (1479, [1])
 (0, [])     (8811, [1, 2])
 (0, [])     (11726, [1, 3])
 (0, [])     (11726, [1, 3])
 (0, [])     (11726, [1, 3])
 (0, [])  …  (11726, [1, 3])
 (0, [])     (11838, [1, 5, 7])
 (0, [])     (13977, [1, 2, 5, 8])
 (0, [])     (13977, [1, 2, 5, 8])
 (0, [])     (13977, [1, 2, 5, 8])
 (0, [])  …  (14607, [1, 5, 8, 9, 11])
 (0, [])     (14607, [1, 5, 8, 9, 11])
 (0, [])     (14714, [1, 5, 8, 9, 13])
 ⋮        ⋱  
 (0, [])     (23965, [1, 5, 11, 13, 14, 19, 24, 26, 33, 37, 38, 56, 65, 82])
 (0, [])     (23965, [1, 5, 11, 13, 14, 19, 24, 26, 33, 37, 38, 56, 65, 82])
 (0, [])  …  (23965, [1, 5, 11, 13, 14, 19, 24, 26, 33, 37, 38, 56, 65, 82])
 (0, [])     (23965, [1, 5, 11, 13, 14, 19, 24, 26, 33, 37, 38, 56, 65, 82])
 (0, [])     (23965, [1, 5, 11, 13, 14, 19, 24, 26, 33, 37, 38, 56, 65, 82])
 (0, [])     (23965, [1, 5, 11, 13, 14, 19, 24, 26, 33, 37, 38, 56, 65, 82])
 (0, [])     (23965, [1, 5, 11, 13, 14, 19, 24, 26, 33, 3

La valeur maximal du sac à dos est de 23965 pour une capacité de 9984.
On retrouve le résultat du TP2, dans ce cas le BRANCH AND BOUND a envoyé la solution exacte.

In [6]:
prix,poids,cap = readKnaptxtInstance("instancesETU/KNAPnewformat/circle/knapPI_16_100_1000_3_-6166.opb.txt")
r = creerTab(prix,poids,cap)

300×1466 Matrix{Tuple{Int64, Vector{Any}}}:
 (0, [])  …  (403, [1])
 (0, [])     (1289, [1, 2])
 (0, [])     (2103, [1, 2, 3])
 (0, [])     (2103, [1, 2, 3])
 (0, [])     (2272, [1, 2, 5])
 (0, [])  …  (2829, [1, 3, 5, 6])
 (0, [])     (2863, [1, 5, 6, 7])
 (0, [])     (2863, [1, 5, 6, 7])
 (0, [])     (2942, [1, 2, 3, 9])
 (0, [])     (2942, [1, 2, 3, 9])
 (0, [])  …  (2942, [1, 2, 3, 9])
 (0, [])     (3049, [6, 9, 10, 12])
 (0, [])     (3049, [6, 9, 10, 12])
 ⋮        ⋱  ⋮
 (0, [])     (7718, [1, 15, 32, 43, 45, 57, 74, 79, 82, 85  …  212, 228, 232, 242, 246, 249, 264, 270, 273, 279])
 (0, [])     (7718, [1, 15, 32, 43, 45, 57, 74, 79, 82, 85  …  212, 228, 232, 242, 246, 249, 264, 270, 273, 279])
 (0, [])  …  (7718, [1, 15, 32, 43, 45, 57, 74, 79, 82, 85  …  212, 228, 232, 242, 246, 249, 264, 270, 273, 279])
 (0, [])     (7718, [1, 15, 32, 43, 45, 57, 74, 79, 82, 85  …  212, 228, 232, 242, 246, 249, 264, 270, 273, 279])
 (0, [])     (7718, [1, 15, 32, 43, 45, 57, 74, 79, 82, 85  …  2

La valeur maximal du sac à dos est de 7904 pour une capacité de 1465.
Ici le branch and bound a renvoyé une valeur inf (6166) et s'est arreté au bout de n itérations.

In [7]:
prix,poids,cap = readKnaptxtInstance("instancesETU/KNAPnewformat/inverse_strongly_correlated/knapPI_4_1000_10000_5_-268720.opb.txt")
r = creerTab(prix,poids,cap)

LoadError: OutOfMemoryError()

L'algorithme peut ne pas fonctionner lorsqu'il y a trop de données, ici le branch an bound trouve son intérêt. L'algotithme implementé est en O(nQ) avec n le nombre d'objets et Q la capcité. 

In [8]:
prix,poids,cap = readKnaptxtInstance("instancesETU/KNAPnewformat/multiple_strongly_correlated/knapPI_14_50_1000_2_-2581.opb.txt")
r = creerTab(prix,poids,cap)

100×983 Matrix{Tuple{Int64, Vector{Any}}}:
 (0, [])  …  (294, [1])
 (0, [])     (1000, [1, 2])
 (0, [])     (1322, [2, 3])
 (0, [])     (1322, [2, 3])
 (0, [])     (1322, [2, 3])
 (0, [])  …  (1580, [1, 5, 6])
 (0, [])     (1580, [1, 5, 6])
 (0, [])     (1580, [1, 5, 6])
 (0, [])     (1580, [1, 5, 6])
 (0, [])     (1580, [1, 5, 6])
 (0, [])  …  (1580, [1, 5, 6])
 (0, [])     (1580, [1, 5, 6])
 (0, [])     (1580, [1, 5, 6])
 ⋮        ⋱  
 (0, [])     (2781, [1, 6, 15, 22, 32, 43, 45, 57, 88])
 (0, [])     (2781, [1, 6, 15, 22, 32, 43, 45, 57, 88])
 (0, [])  …  (2781, [1, 6, 15, 22, 32, 43, 45, 57, 88])
 (0, [])     (2781, [1, 6, 15, 22, 32, 43, 45, 57, 88])
 (0, [])     (2781, [1, 6, 15, 22, 32, 43, 45, 57, 88])
 (0, [])     (2781, [1, 6, 15, 22, 32, 43, 45, 57, 88])
 (0, [])     (2982, [1, 15, 22, 32, 43, 45, 57, 68, 78, 95])
 (0, [])  …  (2982, [1, 15, 22, 32, 43, 45, 57, 68, 78, 95])
 (0, [])     (2982, [1, 15, 22, 32, 43, 45, 57, 68, 78, 95])
 (0, [])     (3181, [15, 22, 30, 32, 43,

La valeur maximal du sac à dos est de 3181 pour une capacité de 982. Ici le branch and bound a renvoyé une valeur inf (2581) et s'est arreté au bout de n itérations.
